Respond with status code 502 if configured feed not yet available

hbruch opened the pull request:
### Summary

This PR changes the 404 to an 503 http status code reply when a requested system feed it not found but it's system_id is configured. 

### Issue

Closes #431 .

### Unit tests

Manually verified.

testower reviewed (commented):
> if system_id is well known and feed is a required one

Do you mean required by the spec? This case is not checked in the implementation though. And wouldn't it be more appropriate to check that it is present in the discovery file?

testower commented on src/main/java/org/entur/lamassu/controller/GBFSV2FeedController.java:
```diff
@@ -98,10 +104,47 @@ public ResponseEntity<Object> getGbfsFeedForProvider(
+    try {
+      GBFS discoveryFile = (GBFS) getFeed(systemId, "gbfs");
+      var feedName = GBFSFeedName.fromValue(feed);
+      List<GBFSFeedName> feedNames =
+        ((GBFS) discoveryFile).getFeedsData()
+          .values()
+          .stream()
+          .map(GBFSFeeds::getFeeds)
+          .flatMap(list -> list.stream())
+          .map(GBFSFeed::getName)
+          .toList();
+
+      return feedNames.contains(feedName);
```
The following adjustment has the advantage that it will return immediately if it finds a match, instead of always going through the whole stream, and collecting at the end before checking for the feed name.

```suggestion
      return discoveryFile.getFeedsData()
                      .values()
                      .stream()
                      .map(GBFSFeeds::getFeeds)
                      .flatMap(Collection::stream)
                      .map(GBFSFeed::getName)
                      .anyMatch(name -> name.equals(feedName));
```

testower commented on src/main/java/org/entur/lamassu/controller/GBFSV2FeedController.java:
```diff
@@ -98,10 +104,47 @@ public ResponseEntity<Object> getGbfsFeedForProvider(
+      return feedNames.contains(feedName);
+    } catch (NoSuchElementException e) {
+      return true;
+    } catch (NullPointerException e) {
```
These catch clauses can be collapsed into one.

testower commented on src/main/java/org/entur/lamassu/controller/GBFSV2FeedController.java:
```diff
@@ -98,10 +104,47 @@ public ResponseEntity<Object> getGbfsFeedForProvider(
+    Returns true, if either the discoveryFile (gbf file) is not yet cached,
+    or the requested feed is published in the discovery file.
+   */
+  protected boolean mightOrShouldFeedExist(String systemId, String feed) {
```
One alternative to returning a boolean here which you have to check at the call site to decide which exception to throw, is to throw the exception directly from within this method. I think that's ok since it's private / local. Could be renamed to "throwIfShouldExist" or something similar.

hbruch replied:
Yes, and I'll move the checking directly into getFeed and throwing the intended exception right there, instead of checking it in the exception handling again.

testower commented:
@hbruch added a few comments

>  I.e. checking the GBFS feeds in mightOrShouldFeedExist and catching exceptions feels a bit clumsy...

Maybe a way to make it feel less clumsy: https://github.com/entur/lamassu/pull/432#discussion_r1579940774 ?

hbruch commented:
I think this PR is good for review. I implemented the same logic for all three updaters and added tests for `GBFSV3FeedUpdateController` and `GBFSV2FeedUpdateController`.
